File: TheBookLore/app/src/main/java/com/penguin/thebooklore/ui/fragment/DashboardFragment.java
package com.penguin.thebooklore.ui.fragment;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.penguin.thebooklore.R;
import com.penguin.thebooklore.databinding.FragmentDashboardBinding;
import com.penguin.thebooklore.model.WebImage;
import com.penguin.thebooklore.ui.adapter.DashboardImagesRecyclerViewAdapter;
import com.penguin.thebooklore.viewmodel.DashboardViewModel;

import java.util.ArrayList;
import java.util.List;

public class DashboardFragment extends Fragment {

    private DashboardViewModel dashboardViewModel;
    private FragmentDashboardBinding binding;
    private Context context;

    private ArrayList<String> imageList = new ArrayList<>();

    private RecyclerView recyclerView;
    private RecyclerView.Adapter mAdapter;
    private RecyclerView.LayoutManager layoutManager;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {

        context = getContext();
        dashboardViewModel = ViewModelProviders.of(this).get(DashboardViewModel.class);
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_dashboard, container, false);
        binding.setLifecycleOwner(getViewLifecycleOwner());
        binding.setViewmodel(dashboardViewModel);

        initObservers();

        recyclerView = (RecyclerView) binding.getRoot().findViewById(R.id.recycler_collection);
        layoutManager = new GridLayoutManager(context, 2);
        recyclerView.setLayoutManager(layoutManager);
        mAdapter = new DashboardImagesRecyclerViewAdapter(context, imageList);
        recyclerView.setAdapter(mAdapter);

        return binding.getRoot();

    }

    private void initObservers() {
        dashboardViewModel.getListImages().observe(this, list -> {
            imageList.clear();
            imageList.addAll(list);
            mAdapter.notifyDataSetChanged();
        });
    }

}

File: TheBookLore/app/src/main/java/com/penguin/thebooklore/ui/fragment/DashboardFragment.kt
package com.penguin.thebooklore.ui.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.GridLayoutManager
import com.penguin.thebooklore.R
import com.penguin.thebooklore.databinding.FragmentDashboardBinding
import com.penguin.thebooklore.model.ArtObject
import com.penguin.thebooklore.model.networkModel.NetworkArtObject
import com.penguin.thebooklore.ui.adapter.DashboardImagesRecyclerViewAdapter
import com.penguin.thebooklore.viewmodel.DashboardViewModel
import java.util.*

class DashboardFragment : Fragment() {

    private val dashboardViewModel: DashboardViewModel by lazy { ViewModelProviders.of(this).get(DashboardViewModel::class.java) }
    private lateinit var binding: FragmentDashboardBinding
    private val artObjectsList = ArrayList<ArtObject>()
    private val layoutManager by lazy { GridLayoutManager(context, 2) }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?, savedInstanceState: Bundle?): View? {

        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_dashboard, container, false)
        binding.lifecycleOwner = viewLifecycleOwner
        binding.viewmodel = dashboardViewModel
        binding.layoutManager = layoutManager
        binding.adapter = DashboardImagesRecyclerViewAdapter(context!!, artObjectsList) { artObject: ArtObject ->  onClickArtObject(artObject)}

        initObservers()

        return binding.root

    }

    private fun onClickArtObject(artObject: ArtObject) {
        Toast.makeText(context, "Open Sesame, ${artObject.title}", Toast.LENGTH_LONG).show()
    }

    private fun initObservers() {
        dashboardViewModel.listArtObjects.observe(this, Observer { list ->
            artObjectsList.clear()
            artObjectsList.addAll(list)
            binding.adapter!!.notifyDataSetChanged()
        })

        dashboardViewModel.isError.observe(this, Observer {exception ->
            exception?.message?.run{
                Toast.makeText(context, this, Toast.LENGTH_LONG).show()
            }
        })
    }

    override fun onDestroyOptionsMenu() {
        super.onDestroyOptionsMenu()
        if (dashboardViewModel.listArtObjects.hasObservers()) dashboardViewModel.listArtObjects.removeObservers(viewLifecycleOwner)
        if (dashboardViewModel.isError.hasObservers()) dashboardViewModel.isError.removeObservers(viewLifecycleOwner)
    }

}